### PR TITLE
Development branch

### PR DESCRIPTION
**Pull request by gitytakahas**
Major updates are followings:

- added MVA met
- generator flags, generator variables (gen boson mass, # of partons)
- variables for the tau polarization measurement
- trigger object name (for trigger matching)
- changed config_MC*.py to run MVA met calculation, and to run on HLT2, instead of HLT, by default

@cgalloni and all who will merge this PR > 
In order to run on MVA MET, you need to add several packages and recompile it again.
Please follow the receipt at https://twiki.cern.ch/twiki/bin/viewauth/CMS/HiggsToTauTauWorking2016#MET

Technically, please do, 

> git cms-addpkg RecoMET/METPUSubtraction
> git cms-addpkg DataFormats/METReco
> git cms-merge-topic rfriese:mvamet80

NB : This will give you some unnecessary packages (which should be due to MVA met author's fault, I believe). For the moment, we have no choice -> I asked authors to remove at some point. 

> mkdir RecoMET/METPUSubtraction/data
> cd RecoMET/METPUSubtraction/data
> wget https://github.com/rfriese/cmssw/raw/MVAMET2_beta_0.6/RecoMET/METPUSubtraction/data/weightfile.root
> scram b clean
> scram b -j8

Please let me know if you see some problem.

### CODE REVIEW

File: Ntuplizer/plugins/METsNtuplizer.cc [hunk 1]
```diff
@@ -231,11 +237,60 @@ void METsNtuplizer::fillBranches( edm::Event const & event, const edm::EventSetu
   event.getByToken(metpuppiInputToken_, METspuppi_ );
 
   for (const pat::MET &met : *METspuppi_) {
-    nBranches_->METpuppi_et.push_back(met.et());
+    nBranches_->METpuppi_et.push_back(met.pt());
```
Comment (cgalloni) [inline Ntuplizer/plugins/METsNtuplizer.cc]: Why did you change into the pt?

File: Ntuplizer/plugins/TriggersNtuplizer.cc [hunk 2]
```diff
@@ -98,27 +98,46 @@ bool TriggersNtuplizer::findTrigger( std::string trigName ){
...
+       trigName.find("HLT_IsoMu19_eta2p1_LooseIsoPFTau20_SingleL1_v") != std::string::npos||
+       trigName.find("HLT_IsoMu19_eta2p1_LooseIsoPFTau20_v") != std::string::npos||
+       trigName.find("HLT_IsoMu21_eta2p1_LooseIsoPFTau20_SingleL1_v") != std::string::npos||
        trigName.find("HLT_IsoMu24_eta2p1") != std::string::npos||
        trigName.find("HLT_IsoMu17_eta2p1") != std::string::npos||
        trigName.find("HLT_IsoMu18_v") != std::string::npos||
        trigName.find("HLT_IsoMu22_v") != std::string::npos||
        trigName.find("HLT_IsoMu27_v") != std::string::npos||
        trigName.find("HLT_IsoMu20_v") != std::string::npos||
+       trigName.find("HLT_IsoMu22_v") != std::string::npos||
```
Comment (cgalloni) [inline Ntuplizer/plugins/TriggersNtuplizer.cc]: this one is already there. Please remove it

**Comment by cgalloni** [comment]
Hi Yuta,
is there a reasone why the configuration files for data have not been changed? also the submit_*py have to be changed, but I can also do it if you tell me which files are needed that are not already in the Ntuplizer directory.
Then could you also update the READ ME file with the newest instruction on how to get the code? 
Thanks,
Camilla

File: Ntuplizer/plugins/METsNtuplizer.cc (hunk 1 above)
Comment (gitytakahas) [inline Ntuplizer/plugins/METsNtuplizer.cc]: For puppi met & PF met, pt() and et() return the same value. 
But for the MVA MET, et() is buggy and pt() is only correct. So I just wanted to make it uniform for all the MET values. 

PS.
missing ET is usually, missing ET "vector", i.e. not the - (scalar sum of ET). So I also prefer pt().

**Comment by gitytakahas** [comment]
Hi Camilla, 

# Just for the record of what we discussed. 

1. For data, we need different configurations for the MVA met, and I still have to check if this works correctly. 

2. Let's check if we need additional corrections (such as missing ET phi, MET recoil corrections). I will work on this from tomorrow afternoon. 

3. I will add READ ME when I checked all above.

So all in all, let's schedule the mass production at the end of this week.

**Comment by gitytakahas** [comment]
Alberto, Camilla and I discussed today about MET recoil correction, and we decided to do this in later stage at SFrame, rather than doing this in Ntuplizer. I added necessary gen flags to do this, as well as MVA METs.

I am planning to add README as well as the configs for data.

**Comment by gitytakahas** [comment]
Added gen particle variables to analyze tau decay product. 
This is indeed redundant, in a sense that it is already included in existing variable, but it should make life much much easier. And yet, these are only filled when there is prompt taus, so it wont't increase Ntuple size.

**Comment by ahinzmann** [comment]
@cgalloni @zucchett  Can you review this? I was not present during the discussion.

File: Ntuplizer/plugins/NtupleBranches.cc [hunk 3]
```diff
@@ -29,17 +29,27 @@ void NtupleBranches::branch( std::map< std::string, bool >& runFlags ){
...
+      tree_->Branch( "genParticle_taudau_eta"	     , &genParticle_taudau_eta        );
+      tree_->Branch( "genParticle_taudau_phi"	     , &genParticle_taudau_phi        );
+      tree_->Branch( "genParticle_taudau_mass"	     , &genParticle_taudau_mass        );
     } //doGenParticles
     
     if ( runFlags["doGenEvent"] ){
       /** generator info */
       tree_->Branch( "lheV_pt"	             , &lheV_pt                ); 
-      tree_->Branch( "lheHT"	             , &lheHT                  ); 
+      tree_->Branch( "lheHt"	             , &lheHt                  );
```
Comment (cgalloni) [inline Ntuplizer/plugins/NtupleBranches.cc]: Is this just to make me change my code? :P

File: Ntuplizer/plugins/METsNtuplizer.cc [hunk 4]
```diff
@@ -231,10 +237,58 @@ void METsNtuplizer::fillBranches( edm::Event const & event, const edm::EventSetu
...
+  event.getByToken(metmvaInputToken_, METsmva_ );
+
+  int nmva = 0;
+
+  for (const pat::MET &met : *METsmva_) {
+    nBranches_->MET_mva_et.push_back(met.pt());
+    nBranches_->MET_mva_phi.push_back(met.phi());
+    nBranches_->MET_mva_cov00.push_back(met.getSignificanceMatrix()(0,0));
+    nBranches_->MET_mva_cov01.push_back(met.getSignificanceMatrix()(0,1));
+    nBranches_->MET_mva_cov10.push_back(met.getSignificanceMatrix()(1,0));
```
Comment (cgalloni) [inline Ntuplizer/plugins/METsNtuplizer.cc]: is element  01 different from element 10  ? this is not the case for the PFMET

File: Ntuplizer/plugins/METsNtuplizer.cc [hunk 5]
```diff
@@ -231,10 +237,58 @@ void METsNtuplizer::fillBranches( edm::Event const & event, const edm::EventSetu
...
+    for(auto name: met.userCandNames()){
+      reco::CandidatePtr aRecoCand = met.userCand(name);
+
+      recoil_pt.push_back(aRecoCand->p4().Pt());
+      recoil_eta.push_back(aRecoCand->p4().Eta());
+      recoil_phi.push_back(aRecoCand->p4().Phi());
+      recoil_pdgId.push_back(aRecoCand->pdgId());
+    }
+
+    nBranches_->MET_mva_recoil_pt.push_back(recoil_pt);
```
Comment (cgalloni) [inline Ntuplizer/plugins/METsNtuplizer.cc]: is this the recoil correction or something else?

File: Ntuplizer/plugins/NtupleBranches.cc [hunk 6]
```diff
@@ -29,17 +29,27 @@ void NtupleBranches::branch( std::map< std::string, bool >& runFlags ){
...
       tree_->Branch( "genParticle_status"    , &genParticle_status     );
+      tree_->Branch( "genParticle_isPrompt"  , &genParticle_isPrompt   );
+      tree_->Branch( "genParticle_isDirectPromptTauDecayProduct"  , &genParticle_isDirectPromptTauDecayProduct);
+      tree_->Branch( "genParticle_isDirectHardProcessTauDecayProductFinalState"  , &genParticle_isDirectHardProcessTauDecayProductFinalState);
+      tree_->Branch( "genParticle_fromHardProcessFinalState"  , &genParticle_fromHardProcessFinalState   );
       tree_->Branch( "genParticle_mother"    , &genParticle_mother     );
       tree_->Branch( "genParticle_nMoth"     , &genParticle_nMoth      );
       tree_->Branch( "genParticle_nDau"	     , &genParticle_nDau       ); 
       tree_->Branch( "genParticle_dau"	     , &genParticle_dau        );
+      tree_->Branch( "genParticle_taudau_pdgId"	     , &genParticle_taudau_pdgId        );
```
Comment (cgalloni) [inline Ntuplizer/plugins/NtupleBranches.cc]: Are the gen tau daughter really needed or can they be done with a loop on the gen particle in SFrame?
I'm a bit worried about the increase of size of the ntuples.

Comment (gitytakahas) [inline Ntuplizer/plugins/NtupleBranches.cc]: This is only filled when we have taus with isPrompt flag is true. So, it won't change the total size at all and this will make our code super easier.

File: Ntuplizer/plugins/METsNtuplizer.cc (hunk 5 above)
Comment (gitytakahas) [inline Ntuplizer/plugins/METsNtuplizer.cc]: This is for the MVA MET (we need to have this in order to check which MVA MET we should take)

File: Ntuplizer/plugins/NtupleBranches.cc (hunk 3 above)
Comment (gitytakahas) [inline Ntuplizer/plugins/NtupleBranches.cc]: will be fixed (maybe, I was just tired ;))

File: Ntuplizer/plugins/METsNtuplizer.cc (hunk 4 above)
Comment (gitytakahas) [inline Ntuplizer/plugins/METsNtuplizer.cc]: these are the same. will be fixed in the next PR

File: Ntuplizer/plugins/NtupleBranches.cc (hunk 3 above)
Comment (cgalloni) [inline Ntuplizer/plugins/NtupleBranches.cc]: ok it's fine also with Ht

File: Ntuplizer/plugins/GenParticlesNtuplizer.cc [hunk 7]
```diff
@@ -39,10 +50,34 @@ GenParticlesNtuplizer::~GenParticlesNtuplizer( void )
...
+
+
       for( unsigned int d=0; d<(*genParticles_)[p].numberOfDaughters(); ++d ){
         vDau.push_back( (*genParticles_)[p].daughter(d)->pdgId() );
-	    nDau++;
+	nDau++;
       }
+
+
+      if(abs((*genParticles_)[p].pdgId())==15 && (*genParticles_)[p].statusFlags().isPrompt()){
```
Comment (cgalloni) [inline Ntuplizer/plugins/GenParticlesNtuplizer.cc]: but like this you don't know which  dauther belong to which tau, don't you?

**Comment by gitytakahas** [comment]
well, I don't think we need to know that, as we only need gen matching. Do you agree ?

**Comment by gitytakahas** [comment]
I added MVAMET flag, so that it can run without any problem (if MVA MET is not solved by Monday).

**Comment by gitytakahas** [comment]
MVA MET problem (in 8_0_20) is fixed (*). Instruction is given in README. 
We can start mass production ! 

@cgalloni 
I tried to compile the list of all the background samples. It is found at

> samples/Summer16

If you are fine, please accept this PR and start mass production. 


(*)
It looks like the default config of the central PATMETProducer that we are cloning to make tauMET producer has been changed recently:

https://github.com/cms-sw/cmssw/commit/a0017a23c6e69cd1c62ed42c17632afd6789662f

It's now pointing at the regular pfCandidates collection found in AOD, not the packed candidates in miniAOD. So as an tentative solution, we had to fix it centrally in the mvaMET config by adding 

> tauMET.srcPFCands =  cms.InputTag("packedPFCandidates")

Instruction is given in README, How to get this latest config.
